docs/proposals/scheduling/multi-podtemplate-support/multiple-pod-template-support.md

```diff
@@ -1,141 +1,129 @@
 
-    // Name of this component
-    Name string `json:"name,omitempty"`
+### 2. API Changes
 
-    // Replicas represents the replica number of the resource's component
-    // +optional
-	Replicas int32 `json:"replicas,omitempty"`
+The main changes of this proposal are to the API definition of the ReplicaRequirements struct. We currently include the replicaCount and replicaRequirements as root level attributes to the ResourceBindingSpec. The limitation here is that we are unable to define unique replicaRequirements in the case that the resource has more than one podTemplate.
 
-    // ReplicaRequirements represents the requirements required by each replica for this component.
-	// +optional
-	ReplicaRequirements *ReplicaRequirements `json:"replicaRequirements,omitempty"`
+To address this, we can move the concept of replicas and replicaRequirements into a struct related to the individual resource's Components.
 
-}
+Each Component will have a Name, the number of Replicas, and corresponding replicaRequirements. These basic fields are necessary to allow the accurate estimator to determine whether all components of the CRD replica will be able to fit on the destination namespace.
 
-// ReplicaRequirements represents the requirements required by each replica.
-type ReplicaRequirements struct {
+The definition of ReplicaRequirements will stay the same - with the drawback that the user will need to define how Karmada interprets the individual components of the CRD. Karmada should also support a default component which will use one of the resource's podTemplates to find requirements.
 
-	// NodeClaim represents the node claim HardNodeAffinity, NodeSelector and Tolerations required by each replica.
-	// +optional
-	NodeClaim *NodeClaim `json:"nodeClaim,omitempty"`
+```go
+type ResourceBindingSpec struct {
+    // ... existing fields ...
+
+    // The total number of replicas scheduled by this resource.
+    TotalReplicas int32 `json:"totalReplicas,omitempty"`
 
-	// ResourceRequest represents the resources required by each replica.
-	// +optional
-	ResourceRequest corev1.ResourceList `json:"resourceRequest,omitempty"`
+    // Defines the requirements of individual components of the resource.
+    // +optional
+    Components []ComponentRequirements `json:"components,omitempty"`
+
+    // ... existing fields ...
+}
 
-	// Namespace represents the resources namespaces
-	// +optional
-	Namespace string `json:"namespace,omitempty"`
+// A component is a unique representation of a resource's replica.
+type ComponentRequirements struct {
+    // Name of this component
+    Name string `json:"name,omitempty"`
 
-	// PriorityClassName represents the components priorityClassName
```

Comment:
Just to clarify, with this change the ComponentRequirements will be added to ResourceBinding spec by default, even if the number of components is 1? I think this would make sense, as the interpreter would default to getReplicas and we can just populate the new ComponentRequirements fields using this single component.

Reply:
Good questions! Actually, I haven't gotten a chance to revisit this part yet; that's the next move. I mean the ResourceBinding changes as well as deal with the relationship between the new ComponentRequirements and the legacy fields.

A draft idea in my mind is that the legacy fields `Replicas` and `ReplicaRequirements` will co-exist with the new `ComponentRequirements` for a while. The legacy field will be populated as usual, but in case of the feature gate is enabled, the new and legacy fields will be both populated. 

The new ComponentRequirements field will replace the legacy fields eventually, once the new feature becomes mature.

Will think and explore more about this.